File: typescript/tests/energy_board.test.ts
// energy_board.test.ts
// Verifies the CircleAI.Energy port: readings-since, total kWh (last-first),
// cost estimate, active outages.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemoryEnergyBoard,
  EnergyDomainContext,
  meterReading,
  energyTariff,
  outage,
} from "../src/energy/index";

describe("InMemoryEnergyBoard", () => {
  it("returns readings since a cutoff ascending", () => {
    const b = new InMemoryEnergyBoard();
    b.record(meterReading("m1", 100, new Date("2026-01-01T00:00:00Z")));
    b.record(meterReading("m1", 150, new Date("2026-01-03T00:00:00Z")));
    b.record(meterReading("m1", 90, new Date("2025-12-30T00:00:00Z"))); // before cutoff
    assert.deepEqual(
      b.readingsFor("m1", new Date("2026-01-01T00:00:00Z")).map((r) => r.kwh),
      [100, 150],
    );
  });

  it("totals kWh as last minus first, 0 when fewer than 2", () => {
    const b = new InMemoryEnergyBoard();
    b.record(meterReading("m1", 100, new Date("2026-01-01T00:00:00Z")));
    assert.equal(b.totalKwhSince("m1", new Date("2026-01-01T00:00:00Z")), 0);
    b.record(meterReading("m1", 175, new Date("2026-01-05T00:00:00Z")));
    assert.equal(b.totalKwhSince("m1", new Date("2026-01-01T00:00:00Z")), 75);
  });

  it("estimates cost at the peak rate; unknown tariff throws", () => {
    const b = new InMemoryEnergyBoard();
    b.record(meterReading("m1", 100, new Date("2026-01-01T00:00:00Z")));
    b.record(meterReading("m1", 200, new Date("2026-01-05T00:00:00Z")));
    b.setTariff(energyTariff("t1", "Standard", 2.5, 1.2, "ZAR"));
    assert.equal(b.getTariff("t1")?.name, "Standard");
    assert.equal(b.estimateCost("m1", "t1", new Date("2026-01-01T00:00:00Z")), 250); // 100 kWh * 2.5
    assert.throws(() => b.estimateCost("m1", "ghost", new Date("2026-01-01T00:00:00Z")), /Unknown tariff ghost/);
  });

  it("lists only ongoing outages (EndUtc == null)", () => {
    const b = new InMemoryEnergyBoard();
    b.logOutage(outage("o1", "North", new Date("2026-01-01T00:00:00Z"), null, "Storm"));
    b.logOutage(outage("o2", "South", new Date("2026-01-01T00:00:00Z"), new Date("2026-01-01T03:00:00Z"), "Fixed"));
    assert.deepEqual(
      b.activeOutages().map((o) => o.outageId),
      ["o1"],
    );
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(EnergyDomainContext.systemPromptSnippet.includes("[DOMAIN: Energy]"));
    assert.deepEqual(EnergyDomainContext.complianceFlags, [
      "Electricity_Act",
      "NERSA",
      "SABS",
      "Municipal_Energy_By_laws",
      "POPIA",
    ]);
    assert.deepEqual(EnergyDomainContext.suggestedTools, ["energy_model", "analytics", "document_editor", "web_search"]);
  });
});

File: go/tests/energy_board_test.go
// energy_board_test.go
//
// Verifies the CircleAI.Energy port (energy_board.go): reading storage +
// since-window ordering, total-kwh (last minus first), tariff set/get, cost
// estimate at peak rate, and active-outage listing.

package circleai_test

import (
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestEnergy_ReadingsAndCost(t *testing.T) {
	b := circleai.NewInMemoryEnergyBoard()
	base := time.Date(2026, 7, 1, 0, 0, 0, 0, time.UTC)
	b.Record(circleai.MeterReading{MeterId: "m1", Kwh: 100, AtUtc: base})
	b.Record(circleai.MeterReading{MeterId: "m1", Kwh: 130, AtUtc: base.Add(24 * time.Hour)})
	b.Record(circleai.MeterReading{MeterId: "m1", Kwh: 175, AtUtc: base.Add(48 * time.Hour)})

	rows := b.ReadingsFor("m1", base)
	if len(rows) != 3 || rows[0].Kwh != 100 || rows[2].Kwh != 175 {
		t.Fatalf("readings-for ordered failed: %+v", rows)
	}
	if kwh := b.TotalKwhSince("m1", base); kwh != 75 {
		t.Fatalf("total kwh = %v, want 75", kwh)
	}

	b.SetTariff(circleai.EnergyTariff{TariffId: "t1", Name: "Std", PeakKwhRate: 2.5, OffPeakKwhRate: 1.0, Currency: "ZAR"})
	if got, ok := b.GetTariff("t1"); !ok || got.Name != "Std" {
		t.Fatalf("get tariff = %+v ok=%v", got, ok)
	}
	// 75 kwh * 2.5 = 187.5.
	cost, err := b.EstimateCost("m1", "t1", base)
	if err != nil {
		t.Fatalf("estimate: %v", err)
	}
	if !cost.Equal(circleai.DecimalFromFloat(187.5)) {
		t.Fatalf("estimate cost = %s, want 187.50", cost.String())
	}
	if _, err := b.EstimateCost("m1", "ghost", base); err == nil {
		t.Fatalf("estimate with unknown tariff must error")
	}
}

func TestEnergy_Outages(t *testing.T) {
	b := circleai.NewInMemoryEnergyBoard()
	now := time.Now().UTC()
	end := now
	b.LogOutage(circleai.Outage{OutageId: "o2", Area: "North", StartUtc: now.Add(-time.Hour)})                   // active
	b.LogOutage(circleai.Outage{OutageId: "o1", Area: "South", StartUtc: now.Add(-2 * time.Hour), EndUtc: &end}) // resolved
	b.LogOutage(circleai.Outage{OutageId: "o3", Area: "East", StartUtc: now.Add(-30 * time.Minute)})             // active

	active := b.ActiveOutages()
	if len(active) != 2 || active[0].OutageId != "o2" || active[1].OutageId != "o3" {
		t.Fatalf("active outages (sorted) failed: %+v", active)
	}
}
